core/src/main/java/io/grpc/internal/MessageFramer.java

```diff
@@ -302,7 +304,11 @@ private void releaseBuffer() {
 
   private void commitToSink(boolean endOfStream, boolean flush) {
     WritableBuffer buf = buffer;
-    buffer = null;
+    // If we have a buffer still and we are not trying to do the last flush on a close() call,
+    // try and reuse the remainder of the buffer.
+    if (buf != null && !isClosed()) {
```

Comment:
@louiscryan 

Don't think that works. Note that not only the backing memory is pooled / reused, but also the `ByteBuf` object itself. How do you distinquish if a `ByteBuf` object with `refCnt() > 0` is still that same object with the same backing memory or if it has meanwhile been claimed by some other call side? Even if additionally to the ref count we compared `ByteBuf.memoryAddress`, it wouldn't be safe as 1) we can't compare `memoryAddress` + `refCnt` and increment `refCnt` atomically and 2) by chance another call side might have gotten the same `ByteBuf` object with the same backing memory and we can't distinquish that case.